mm2src/kdf_walletconnect/src/connection_handler.rs

```diff
@@ -0,0 +1,137 @@
+}
+
+impl Handler {
+    pub fn new(
+        name: &'static str,
+        msg_sender: UnboundedSender<PublishedMessage>,
+        conn_live_sender: UnboundedSender<Option<String>>,
+    ) -> Self {
+        Self {
+            name,
+            msg_sender,
+            conn_live_sender,
+        }
+    }
+}
+
+impl ConnectionHandler for Handler {
+    fn connected(&mut self) {
+        debug!("[{}] connection to WalletConnect relay server successful", self.name);
+    }
+
+    fn disconnected(&mut self, frame: Option<CloseFrame<'static>>) {
+        debug!("[{}] connection closed: frame={frame:?}", self.name);
+
+        if let Err(e) = self.conn_live_sender.unbounded_send(frame.map(|f| f.to_string())) {
+            error!("[{}] failed to send to the receiver: {e}", self.name);
+        }
+    }
+
+    fn message_received(&mut self, message: PublishedMessage) {
+        debug!(
+            "[{}] inbound message: message_id={} topic={} tag={} message={}",
+            self.name, message.message_id, message.topic, message.tag, message.message,
+        );
+
+        if let Err(e) = self.msg_sender.unbounded_send(message) {
+            error!("[{}] failed to send to the receiver: {e}", self.name);
+        }
+    }
+
+    fn inbound_error(&mut self, error: ClientError) {
+        debug!("[{}] inbound error: {error}", self.name);
+        if let Err(e) = self.conn_live_sender.unbounded_send(Some(error.to_string())) {
+            error!("[{}] failed to send to the receiver: {e}", self.name);
+        }
+    }
+
+    fn outbound_error(&mut self, error: ClientError) {
+        debug!("[{}] outbound error: {error}", self.name);
+        if let Err(e) = self.conn_live_sender.unbounded_send(Some(error.to_string())) {
+            error!("[{}] failed to send to the receiver: {e}", self.name);
+        }
+    }
+}
+
+/// Establishes initial connection to WalletConnect relay server with linear retry mechanism.
+/// Uses increasing delay between retry attempts starting from INITIAL_RETRY_SECS.
+/// After successful connection, attempts to restore previous session state from storage.
+pub(crate) async fn spawn_connection_initialization(
+    wc: Arc<WalletConnectCtxImpl>,
+    connection_live_rx: UnboundedReceiver<Option<String>>,
+) {
+    info!("Initializing WalletConnect connection");
+    let mut retry_count = 0;
+    let mut retry_secs = INITIAL_RETRY_SECS;
+
+    while let Err(err) = wc.connect_client().await {
+        retry_count += 1;
+        error!(
+            "Error during initial connection attempt {}: {:?}. Retrying in {retry_secs} seconds...",
+            retry_count, err
+        );
+        Timer::sleep(retry_secs).await;
+        retry_secs += RETRY_INCREMENT;
+    }
+
+    // Initialize storage
+    if let Err(err) = wc.session_manager.storage().init().await {
+        error!("Unable to initialize WalletConnect persistent storage: {err:?}. Only inmemory storage will be utilized for this Session.");
+    };
```

Comment:
do we really utilize any in-memory storage on `init` failure? i can't find that anywhere.

Reply:
aha i see. the error message made me thing we opened an `InMemorySqlDB` and used it instead.
what about saying: `Unable to initialize WalletConnect persistent storage: {err:?}. Proceeding without a storage (changes will not persist!).`

Reply:
yeah, I think we should panic if there's a problem with storage on start

Reply:
I thought panic might be the best case here so I opted for aborting all WC context spawned futures on failure while initializing.